src/main/java/org/verdictdb/coordinator/AsyncAggExecutionRulerByConverge.java

```diff
@@ -0,0 +1,126 @@
+
+  AsyncAggExecutionRulerByConverge(VerdictResultStream resultStream, Coordinator runningCoordinator) {
+    this.resultStream = resultStream;
+    this.runningCoordinator = runningCoordinator;
+  }
+
+  public void setValueError(Double valueError) {
+    this.valueError = valueError;
+  }
+
+  public void setColumnError(Double columnError) {
+    this.columnError = columnError;
+  }
+
+  /**
+   * fetch the answer from stream until converge
+   * @return the accurate answer
+   */
+  @Override
+  public VerdictSingleResult fetchAnswerUntilAccurate() {
+    try {
+      while (resultStream.hasNext()) {
+        answers.add(resultStream.next());
+        if (checkConverge()) {
+          log.debug("Break condition has reached.");
+          break;
+        }
+      }
+    } catch (RuntimeException e) {
+      throw e;
+    }
+    resultStream.close();
+    log.debug("Aborts an ExecutionContext: " + this);
+    if (runningCoordinator != null) {
+      Coordinator c = runningCoordinator;
+      runningCoordinator = null;
+      c.abort();
+    }
+    return answers.get(answers.size()-1);
+  }
+
+  private boolean checkConverge() {
+    HashMap<List<Object>, List<Object>> newAggregatedMap = new HashMap<>();
+    VerdictSingleResult currentAnswer = answers.get(answers.size()-1);
+    while (currentAnswer.next()) {
+      List<Object> aggregatedValues = new ArrayList<>();
+      List<Object> nonAggregatedValues = new ArrayList<>();
+      for (int i=0;i<currentAnswer.getColumnCount();i++) {
+        if (currentAnswer.getMetaData().isAggregate.get(i)) {
+          aggregatedValues.add(currentAnswer.getValue(i));
+        } else {
+          nonAggregatedValues.add(currentAnswer.getValue(i));
+        }
+      }
+      newAggregatedMap.put(nonAggregatedValues, aggregatedValues);
+    }
+    aggregatedMap = newAggregatedMap;
+    currentAnswer.rewind();
+
+    if (answers.size()==1) {
+      return false;
+    }
+    VerdictSingleResult previousAnswer = answers.get(answers.size()-2);
+
+
+    // check if #column is converged
+    if (currentAnswer.getRowCount() < previousAnswer.getRowCount()*(1-columnError)
+        || currentAnswer.getRowCount() > previousAnswer.getRowCount()*(1+columnError)) {
+      return false;
+    }
+
+    Boolean isValueConverged = true;
+    for (List<Object> nonAggregatedValues:newAggregatedMap.keySet()) {
+      if (isValueConverged && aggregatedMap.containsKey(nonAggregatedValues)) {
+        List<Object> prevAggregatedValues = aggregatedMap.get(nonAggregatedValues);
+        List<Object> aggregatedValues = newAggregatedMap.get(nonAggregatedValues);
+        for (Object v:aggregatedValues) {
+          int idx = aggregatedValues.indexOf(v);
+          double newValue, oldValue;
+          if (v instanceof BigDecimal) {
+            newValue = ((BigDecimal) v).doubleValue();
+            oldValue = ((BigDecimal) prevAggregatedValues.get(idx)).doubleValue();
+          } else {
+            newValue = (double) v;
```

Comment:
Is this casting safe for integer?

Reply:
I don't think this comment has been addressed. Otherwise, please leave a comment.